File: 2024/42/button.rb
#!/usr/bin/env ruby

# 2024 Day 21 Part 2 / Puzzle 2024/42

# Something about the shortest button string to force a robot (indirect through
# 2 robots and yourself) to enter a given code successfully.

# 7 8 9
# 4 5 6
# 1 2 3
#   0 A
#
# and
#
#   ^ A
# < v >
#
# are the button layouts. The robot always starts on 'A'.
#
# Written mostly in a cabin in a West Virginia state park w/ no internet
# access.
class Puzzle
  def initialize()
    @codes = [ ]
    @row_pos = { }
    @col_pos = { }
    @move_cache = { }
    @code_cache = { }

    %w[7 8 9].each { |i| @row_pos[i] = 0 }
    %w[4 5 6].each { |i| @row_pos[i] = 1 }
    %w[1 2 3].each { |i| @row_pos[i] = 2 }
    %w[  0 A].each { |i| @row_pos[i] = 3 }

    %w[7 4 1  ].each { |i| @col_pos[i] = 0 }
    %w[8 5 2 0].each { |i| @col_pos[i] = 1 }
    %w[9 6 3 A].each { |i| @col_pos[i] = 2 }

    # We can use same coord space for the num pad and dir pad
    # if we orient around 'A' always being in the same spot
    %w[< v >].each { |i| @row_pos[i] = 4 }
    @row_pos['^'] = 3 # next to A
    @col_pos['<'] = 0
    @col_pos['^'] = 1
    @col_pos['v'] = 1
    @col_pos['>'] = 2
  end

  def decode(line)
    @codes.push line.chomp
  end

  def clear_cache
    @move_cache = { }
    @code_cache = { }
  end

  def codes
    @codes
  end

  # Find keystroke sequences to move us (dx, dy) starting from (row, col)
  def resolve_to_zero(dx, dy, row, col)
    res = []
    if dx == 0 && dy == 0
      return ''
    end

    # ever pointing at row 3, col 0 (the blank space left of '^' and below '1')
    # will cause the robot to instantly panic.

    # we can get shortest strings by going the same direction repeatedly, so
    # only try the two cardinal directions we can go (accounting for the poison
    # cell above)
    x_str = (dx < 0) ? '<' : '>'
    x_str = x_str * dx.abs

    y_str = (dy < 0) ? '^' : 'v'
    y_str = y_str * dy.abs

    # needed for the .between? call, which requires first param be <= second
    minc = [col, col + dx].min
    maxc = [col, col + dx].max

    if dx != 0
      if dy == 0 || col == 0 || (!0.between?(minc, maxc) || row != 3)
        # dy == 0: if we're only moving in the current row we won't be asked to touch
        # the wrong gap.
        # col == 0: If col == 0 we're moving away, and we should do so first.
        # 0.between?: Is it possible to hit panic area with wrong x/y seq? If
        # so, ensure we're not already in the wrong row (if we are, the check
        # below will move us then).
        res.push (x_str + y_str)
      end
    end

    minr = [row, row + dy].min
    maxr = [row, row + dy].max

    if dy != 0
      # Similar to the logic of the col-first sequence above
      if dx == 0 || row == 3 || (!3.between?(minr, maxr) || col != 0)
        res.push (y_str + x_str)
      end
    end

    if (dx != 0 || dy != 0) && res.empty?
      throw "Empty result when a result is required! row=#{row}, col=#{col}, dx=#{dx}, dy=#{dy}"
    end

    return res
  end

  def row_of(ch)
    throw "row: #{ch}" unless @row_pos.has_key?(ch)
    return @row_pos[ch]
  end

  def col_of(ch)
    thcol "col: #{ch}" unless @col_pos.has_key?(ch)
    return @col_pos[ch]
  end

  def shortest_move_commands(start, stop)
    # Rather than Djikstra or whatever, just take advantage of the fact we know
    # we have to move over the specific number of rows and columns

    if start == stop
      # Hey look we're already there
      return ["A"]
    end

    key = "#{start}#{stop}"
    if @move_cache.has_key?(key)
      return @move_cache[key]
    end

    start_row = row_of(start)
    stop_row  = row_of(stop)
    start_col = col_of(start)
    stop_col  = col_of(stop)

    dx = stop_col - start_col
    dy = stop_row - start_row

    opts = resolve_to_zero(dx, dy, start_row, start_col)
    opts.each { |opt| opt.concat('A') }

    groups = opts.group_by { |x| x.length }
    min_group = groups.keys.min

    opts = groups[min_group]
    @move_cache[key] = opts
    return opts
  end

  def do_code_step(cur_ch, level, next_ch, rest)
    # if level = 0, we no longer recursively sub-expand our strings to find
    # their lengths but instead return length directly

    res = shortest_move_commands(cur_ch, next_ch)

    if level == 0
      min_len = res.map { |x| x.length }.min

      if rest.empty?
        # base case
        return min_len
      else
        # sum across
        remainder = do_code_step(next_ch, level, rest[0], rest[1..])
        return min_len + remainder
      end
    end

    key = [level, '/', cur_ch, next_ch, rest].join
    if @code_cache.has_key?(key)
      return @code_cache[key]
    end

    # level != 0
    # we need to drill down on the sub-components broken up by each 'A' press
    # and find the min-length of each possible segment, to find the best
    # segment we can report as an overall possibility.  The segments that sum
    # to the smallest value is the overall minimum length of the overall
    # possibility
    possible_lens = res.map { |overall_seg|
        do_code_step('A', level - 1, overall_seg[0], overall_seg[1..])
    }

    remainder = 0
    if !rest.empty?
      remainder = do_code_step(next_ch, level, rest[0], rest[1..])
    end

    @code_cache[key] = possible_lens.min + remainder
    return @code_cache[key]
  end
end

p = Puzzle.new()

File.foreach(ARGV.shift || '../41/input') { |line|
  p.decode(line.chomp)
}

puts "Puzzle codes: #{p.codes}"

sum = 0

p.codes.each { |code|
  min_len = p.do_code_step('A', 25, code[0], code[1..])
  puts "#{code} min len? #{min_len}"
  sum += code[0..3].to_i * min_len
}

puts "Final sum: #{sum}"

File: 2024/42/button.rs
use std::collections::HashMap;
use std::env;
use std::fs;

// Advent of Code: 2024 day 21, part 2

type Pos = (usize, usize);
type PosSet = HashMap<u8, Pos>;
type MoveCache = HashMap<String, Vec<String>>;
type CodeCache = HashMap<String, usize>;

struct Puzzle {
    pos_set: PosSet,
    move_cache: MoveCache,
    code_cache: CodeCache,
}

impl Puzzle {
    fn resolve_to_zero(&self, dx: i32, dy: i32, start: Pos) -> Vec<String> {
        let mut res: Vec<String> = vec![];
        let (row, col) = start;

        if dx == 0 && dy == 0 {
            return res;
        }

        // See paired Ruby script for more comments about why
        let mut x_str = if dx < 0 { "<" } else { ">" };
        let tmp = &x_str.repeat(dx.abs() as usize);
        x_str = tmp;

        let mut y_str = if dy < 0 { "^" } else { "v" };
        let tmp2 = &y_str.repeat(dy.abs() as usize);
        y_str = tmp2;

        let minc = col.min((col as i32 + dx) as usize);

        if dx != 0 && (dy == 0 || col == 0 || minc > 0 || row != 3) {
            res.push(x_str.to_owned() + y_str);
        }

        let minr = row.min((row as i32 + dy) as usize);
        let maxr = row.max((row as i32 + dy) as usize);

        if dy != 0 && (dx == 0 || row == 3 || (minr > 3 || maxr < 3) || col != 0) {
            res.push(y_str.to_owned() + x_str);
        }

        if res.len() == 0 {
            panic!("Empty result!");
        }

        return res;
    }

    fn shortest_move_commands(&mut self, start: u8, stop: u8)
        -> Vec<String>
    {
        let mut res: Vec<String> = vec![];

        if start == stop {
            res.push("A".to_string());
            return res;
        }

        let key = start.to_string() + &stop.to_string();
        if self.move_cache.contains_key(&key) {
            return self.move_cache[&key].clone();
        }

        // I messed this up in transcription from Ruby in gen_pos_set, so
        // fixup by swapping col/row here (but only here)
        let (start_col, start_row) = self.pos_set[&start];
        let (stop_col, stop_row)   = self.pos_set[&stop];
        let dx = stop_col as i32 - start_col as i32;
        let dy = stop_row as i32 - start_row as i32;

        let mut opts = self.resolve_to_zero(dx, dy, (start_row, start_col));
        for o in opts.iter_mut() {
            o.push('A');
        }

        let min_size = opts.iter().map(String::len).min().unwrap();
        opts.retain(|x| x.len() == min_size);

        self.move_cache.insert(key, opts.clone());
        return opts;
    }

    fn do_code_step(&mut self, cur_ch: u8, level: usize, next_ch: u8, rest: &str) -> usize {
        // At level 0 we no longer recursively sub-expand strings to find their
        // min lengths, but instead return directly. But we still must account for
        // the rest of the string either way.

        let res = self.shortest_move_commands(cur_ch, next_ch);

        let remainder = if rest.is_empty() {
            0
        } else {
            let bs = rest.as_bytes();
            self.do_code_step(next_ch, level, bs[0], &rest[1..])
        };

        let key = level.to_string() + "/" + &cur_ch.to_string() + &next_ch.to_string() + rest;
        if self.code_cache.contains_key(&key) {
            return self.code_cache[&key];
        }

        let min_len = match level {
            0 => res.iter().map(String::len).min().unwrap(),
            _ => res.iter().map (|x| {
                     let our_bs = x.as_bytes();
                     self.do_code_step(b'A', level - 1, our_bs[0], &x[1..])
                 }).min().unwrap()
        };

        let result = min_len + remainder;
        self.code_cache.insert(key, result);

        return result;
    }

}

fn gen_pos_set() -> PosSet {
    let mut pos = PosSet::new();

    // 7 8 9  <-- layout
    // 4 5 6
    // 1 2 3
    //   0 A
    //
    // and
    //
    //   ^ A  <-- note this 'A' *OVERLAPs* the one above
    // < v >

    // these were inadvertently col,row order when they should have been
    // row,col order.  Fixed in the one consumer of this function.
    pos.insert(b'7', (0, 0));
    pos.insert(b'8', (1, 0));
    pos.insert(b'9', (2, 0));

    pos.insert(b'4', (0, 1));
    pos.insert(b'5', (1, 1));
    pos.insert(b'6', (2, 1));

    pos.insert(b'1', (0, 2));
    pos.insert(b'2', (1, 2));
    pos.insert(b'3', (2, 2));

    pos.insert(b'^', (1, 3)); // overlap
    pos.insert(b'0', (1, 3));
    pos.insert(b'A', (2, 3));

    pos.insert(b'<', (0, 4));
    pos.insert(b'v', (1, 4));
    pos.insert(b'>', (2, 4));

    return pos;
}

fn main() {
    let default_filename: &'static str = "../41/input";
    let args: Vec<String> = env::args().collect();
    let in_file: String = match args.iter().filter(|a| !a.starts_with("-")).count() {
        // No args provided save argv[0]
        1 => String::from(default_filename),
        _ => args[args.len() - 1].clone(),
    };

    let lines = fs::read_to_string(&in_file)
        .expect("Should have been able to read the file")
        .lines()
        .map(String::from)
        .collect::<Vec<_>>();

    let pos_set = gen_pos_set();
    let move_cache = MoveCache::new();
    let code_cache = CodeCache::new();
    let mut p = Puzzle { pos_set, move_cache, code_cache };

    let mut sum = 0;

    for code in lines.iter() {
        let bs = code.as_bytes();
        let min_len = p.do_code_step(b'A', 25, bs[0], &code[1..]);
        println!("{} min len? {}", code, min_len);

        let code_num = &code[0..3].parse::<usize>().unwrap();
        sum = sum + code_num * min_len;
    }

    println!("Sum is {}", sum);
}
